### DISCUSSION TITLE
Any reason _not_ to use repo v3 rev as subscribeRepos sequence number?

### DISCUSSION THREAD

**Discussion by snarfed** [bluesky-social/atproto#1607] [Protocol (atproto)]
Hi all! So in [repo v3](https://atproto.com/blog/repo-sync-update), `rev` is a [TID](https://atproto.com/specs/record-key#record-key-type-tid), ie a base32-encoded integer, and it has broadly the same semantics and requirements as the `subscribeRepos` [sequence number](https://atproto.com/specs/event-stream#sequence-numbers). I like the separation between the two in the specs, esp since there are other subscription methods, eg `subscribeLabels`. As an implementation detail though, can you think of any reason not to reuse a commit's `rev` as its sequence number?

(As a hint, https://atproto.com/specs/event-stream#sequence-numbers says "For example, they might be timestamps," so I expect the answer is no. 😁)

**Comment by DavidBuchanan314** [comment]
So I'm currently using a microsecond-resolution timestamp counter as my sequence number, in picopds. It works quite nicely and it's convenient to implement, but I can think of a couple of things that *might* go wrong:

- It's not strongly guaranteed to be monotonic - for example, having the system clock drift forwards, followed by an NTP re-sync bringing it back to the correct time. (but I suppose this would also be an issue for your `rev` values in the first place)

- I forgot what the other thing was... If I remember, I'll come back and update this

**Reply by snarfed** [reply]
Definitely! Afaik local clocks aren't always guaranteed to be monotonic even on their own, even before considering explicit changes like NTP etc make.

Fortunately both `rev` and `seq` _are_ required to be monotonically increasing, so if we manage to provide that guarantee for one, we get it "for free" if we reuse it for the other.

**Reply by snarfed** [reply]
One interesting wrinkle with `rev` is that it's a TID, which isn't monotonically increasing globally or even per-repo due to the clock id. From https://atproto.com/specs/record-key#record-key-type-tid :

> TID generators should generate a random clock identifier number, chosen to avoid collisions as much as possible (for example, between multiple worker instances of a PDS service cluster). A local clock can be used to generate the timestamp itself. Care should be taken to ensure the TID output stream is monotonically increasing and never repeats, even if multiple TIDs are generated in the same microsecond, or during "clock smear" or clock synchronization incidents.

If I read that right, the "monotonically increasing" part is scoped to a given clock id, ie (implicitly) machine. `rev`, on the other hand, is scoped to a given _repo_. So if your PDS runs on more than one machine, you can't necessarily your per-machine TID generator for record rkeys to also generate `rev`s, since the clock id will prevent it from being monotonic.

(I think I have that right...?)

**Reply by DavidBuchanan314** [reply]
It doesn't answer the question, but this thread might be handy for context: https://bsky.app/profile/retr0.id/post/3k5pvuta5pb2v

I'll reproduce it here for link-rot-avoidance purposes:

me, quoting a thread announcing repo v3:

> If repo revisions are monotonic TIDs, what are we supposed to do about the 10-bit "clock identifier" part of the TID?

> There are 3 options I can think of:
> 1. require it be all zeroes
> 2. allow arbitrary values but pretend they're zeroes for comparison purposes
> 3. allow arbitrary values, compare by magnitude (i.e. equivalent to lexicographic comparison of raw TID strings - nice and simple)

> This also raises the question of "what do you do when you have two unique commits with the same TID?"

Bryan Newbold:

> compare by overall TID.
> if you have a commit for a did, and discover a new different commit with same did+tid, reject/ignore it. if you discover two different did+tid commits at the same time, something is fishy/hostile. i guess we should define something, but probably mark repo as suss

**Reply by snarfed** [reply]
Yes! It seems like the original goal for TIDs was collision resistance, since they're record keys. That makes sense. However, they aren't monotonically increasing across machines, so it's surprising to see them reused here for a use case that needs that, at least per repo.

**Reply by snarfed** [reply]
For now, to get my own code to repo v3, I guess I can implement a new, storage-based TID generator that provides per-repo monotonicity across machines. I'll probably have it always set clock id to 0.

**Reply by bnewbold** [reply]
clock identifier: we aren't really using it to prevent collisions as originally described

The recommendation/assumption for TIDs is that they be tracked by a "TID generator clock" which remembers the last value it gave out as state. When generating a new TID, take a timestamp. If it is lower or equal to the previous TID output, then ignore the timestamp and increment the last value by one and output that. This ensures monotonic output. It assumes the clock never gets too out of whack (like a 100 years). If you wanted to keep the clock ID part of the identifier clean, you'd increment the microseconds part instead of incrementing the overall TID value I guess. Often this is done as a thread-safe sequencing service, or even a microservice.

The above is a common pattern, but for atproto we don't require all this and are often chill. TIDs in rkey or commit `rev` are per-repo, and as a PDS you usually have the previous `rev` to compare against. The rkey values in records really don't matter that much (at least for now), and it doesn't matter if `rev` TID is equal, greater, lesser than any rkey TIDs.

The event stream sequence numbers are PDS-scoped (not repo-scoped). If you are doing a single repo PDS I guess it would be easy in some ways to use the `rev`? But otherwise I wouldn't bother. The model for repo event stream sequence numbers is definitely Kafka offsets, which start from zero and go up monotonically.

We may add the concept of a event stream sequence "epoch"/"era" to handle resets/resharding; and sharding of the stream, possibly using multiple (many?) parallel sequence numbers.

**Reply by snarfed** [reply]
Thanks! "as a PDS you usually have the previous `rev` to compare against" is a useful hint here.

Event stream sequencers (eg `subscribeRepos` connection handlers) are already single-homed (ie served from a single nodes) due to the websocket connection, so requiring them to serially generate sequence numbers PDS-wide seems fine, and easy to keep contained and out of the way.

There isn't a similar inherent constraint that repos are single-homed, though, and most straightforward distributed PDS designs probably wouldn't pin repos to specific nodes. So, they'd need some other place to synchronize generating `rev`s across nodes.

We have to linearize commits per repo anyway though, and loading `prev` in that process to check that the new `rev` is higher is easy and makes sense. Thanks!

**Comment by DavidBuchanan314** [comment]
Another reason more specific to using the repo `rev` is that revs are only required to be monotonic per-repo (it would be legal for two repos on the same PDS to have the same rev, iiuc). Perhaps you have some logic that would prevent this from being the case, but something to be aware of at least.

**Reply by bnewbold** [reply]
Yup. I would recommend against using the `rev` as sequence number for this reason.

It is totally valid to think of using `rev` for the sequence number: it is weird to have basically sequence numbers per repo, and then *again* for the event stream. But the scope of ordering is different.
